### ISSUE TITLE
[Proposal]: Fluent Bit Extended Buffering Options for Output Plugins

### ISSUE THREAD

**Issue by JeffLuoo** [fluent/fluent-bit#2408]
## Objective
This proposal will show some configuration parameters we propose to add to the Fluent Bit storage layer for enhancing the offline buffering for logs.

## Background
Buffer & Storage is essential for log processing as it handles the backpressure and ensures the data integrity. This becomes more important especially for edge use cases (unreliable networking environment etc.).

Fluent Bit already provides a secondary buffering mechanism based in the file system. It can be enabled per-ingestor plugin, so every record and chunk registered keeps a copy in the file system until it is fully delivered to the destination. However there are some limitations (details in the next section) in the current configuration which we want to address in this proposal.

## Critical User Journey (CUJ)
As a Fluent Bit user:

- CUJ 1: I want my logs to be buffered locally if my network goes down for a couple of hours, so that I will not lose logs. 

- CUJ 2: I want to configure the offline buffering behavior, e.g. “keep offline logs with a maximum size of 1 GB and created in the last 2 hours (time)”.

- CUJ 3: I want to prevent the disk becoming full by buffering input chunks without limit so other applications cannot use the disk. 

- CUJ 4:  I want to discard the data if the logs are older than X hours because the remote API won't accept it.

CUJ1 can be fulfilled today (for stackdriver output plugin) by setting Retry_Limit False. However there are no configuration options for fulfilling CUJ2/3/4. 

## Proposal
Configuration parameters below are designed in the output plugins.
### Option 1: storage.total_limit_size

- Description: The size limitation of this output plugin instance can buffer in the filesystem. Need to be used together with input plugins configured with storage.type as filesystem. Will drop the oldest log data chunks when offline buffer size exceeds this limit. 
- Explanation: for limiting size of offline buffer. 
- Default: -1 (unlimited)

### Option 2: storage.ignore_older

- Description: Drop log data chunks with creation time older than this time in seconds. Supports m,h,d (minutes, hours, days) syntax. 
- Explanation: for limiting time of offline buffer.
- Default: Nil

### Option 3: backoff.retry_max_interval

- Description: Cap the maximum retry time scheduled by the flb_scheduler in seconds. Supports m,h,d (minutes, hours, days) syntax.

- Explanation:

    - Now the next retry time is scheduled by the scheduler using a Truncated Exponential Backoff algorithm - essentially an exponential backoff with a maximum. 
    - Currently the maximum is not configurable - 2000 second. 
    - If you set retry_max_interval to 1000 seconds then the maximum retry time will be capped at 1000 seconds (instead of growing infinitely).

- Default: 2000 (current cap)

## Configuration Example
```
[SERVICE]
    flush                     1
    log_Level                 info
    storage.path              /var/log/flb-storage/
    storage.sync              normal
    storage.checksum          off
    storage.backlog.mem_limit 5M

[INPUT]
    name          cpu
    storage.type  filesystem

[INPUT]
    name          mem
    storage.type  filesystem

[OUTPUT]
    Name stackdriver
    Match *
    Retry_Limit False
    storage.total_limit_size  2G
    storage.ignore_older      4h
    backoff.retry_max_interval 10m
```

Please feel free to comment and provide feedback on the proposal. Thanks in advance for any suggestion and idea :-)

**Comment by JeffLuoo** [comment]
Hi @edsiper Could you please review this proposal and provide some feedback?

**Comment by edsiper** [comment]
This was implemented by @JeffLuoo in v1.6 release:

extended docs here: https://docs.fluentbit.io/manual/administration/buffering-and-storage#chunks-memory-filesystem-and-backpressure
